modules/markdown/assets/js/components/markdown-live-preview.jsx

```diff
@@ -0,0 +1,193 @@
+	};
+	// Adds `#`s to the beginning of the heading content
+	markdownIt.renderer.rules.heading_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		const inline_token = tokens[ idx + 1 ];
+		const text_token = inline_token.children[ 0 ];
+		if ( text_token ) {
+			text_token.content = ` ${ text_token.content }`;
+		}
+		return `<${ token.tag }>${ token.markup }`;
+	};
+};
+
+const renderHTML = function( source ) {
+	if ( source ) {
+		return markdownIt.render( source )
+		// The MarkdownIt renderer adds new lines after each paragraph/heading
+		// tag. This affects the restoration of the caret position, so we need
+		// to remove them
+			.split( /\n/ ).join( '' )
+			.trim();
+	}
+};
+
+const stripTrailingNewLines = function( text ) {
+	if ( text ) {
+		return text.replace( /\n{1,2}$/, '' );
+	}
+};
+
+const triggerOnChange = function( evt, source ) {
+	if ( this.props.onChange ) {
+		// Clone event with Object.assign to avoid
+		// "Cannot assign to read only property 'target' of object"
+		evt = Object.assign( {}, evt, {
+			target: {
+				value: source
+			}
+		} );
+		this.props.onChange( evt );
+	}
+};
+
+const sourceIsEmpty = function( source ) {
+	return ! source || '' === source.trim();
+};
+
+const ignoreLastInput = function( source ) {
+	return endsWith( source, ' ' ) || endsWith( source, ' ' );
+};
+
+const emptyState = '<p></p>';
+
+const emitChange = function( evt ) {
+	if ( ! this.htmlEl ) {
+		return true;
+	}
+
+	// We need to delete the last two new lines added by the browser for the
+	// last node in the component's content. This behaviour interferes with
+	// headings parsing.
+	const source = stripTrailingNewLines( this.htmlEl.innerText );
+
+	triggerOnChange.call( this, evt, source );
+
+	// if there's no source, we don't need to parse anything
+	if ( sourceIsEmpty( source ) ) {
+		this.htmlEl.innerHTML = emptyState;
+		return true;
+	}
+
+	// commonmark doesn't allow trailing spaces in paragraphs, so if we've
+	// added a space, calling markdownIt now will remove it
+	if ( ignoreLastInput( source ) ) {
+		return true;
+	}
+
+	this.setState( { restoreCaretPosition: null } );
+
+	if ( source ) {
+		const html = renderHTML( source );
+
+		this.setState( {
+			html
+		} );
+	}
+
+	return true;
+};
+
+export default class MarkdownLivePreview extends React.Component {
+
+	constructor( props ) {
+		super();
+
+		const { source } = props;
+
+		setupMarkdownParser();
+
+		this.state = {
+			html: source ? renderHTML( source ) : emptyState,
+		};
+	}
+
+	render() {
+		const { ...props } = this.props;
+
+		return createElement(
+			'div',
+			{
+				...props,
+				ref: ( e ) => this.htmlEl = e,
+				onInput: emitChange.bind( this ),
+				onBlur: emitChange.bind( this ),
+				contentEditable: ! this.props.disabled,
+				dangerouslySetInnerHTML: { __html: this.state.html }
+			},
+			this.props.children
+		);
+	}
+
+	shouldComponentUpdate( nextProps, nextState ) {
```

Comment:
Thinking of the future, `getSnapshotBeforeUpdate` will probably be more apt for the caret position saving.